server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java

```diff
@@ -49,188 +51,313 @@
+    for (int numAssigned = 0; numAssigned < numToAssign; numAssigned++) {
+      if (!throttler.canCreateReplicant(tier)) {
+        return numAssigned;
       }
 
-      final ServerHolder holder = strategy.findNewSegmentHomeReplicator(segment, serverHolderList);
-
+      // Retrieves from cache if available
+      ServerHolder holder = strategyCache.remove(tier);
+      // Does strategy call if not in cache
       if (holder == null) {
-        log.warn(
-            "Not enough [%s] servers or node capacity to assign segment[%s]! Expected Replicants[%d]",
-            tier,
-            segment.getIdentifier(),
-            expectedReplicantsInTier
-        );
-        break;
+        holder = params.getBalancerStrategy().findNewSegmentHomeReplicator(segment, holders);
       }
 
-      if (replicate) {
-        replicationManager.registerReplicantCreation(
-            tier, segment.getIdentifier(), holder.getServer().getHost()
+      if (holder == null) {
+        log.warn(
+            "No available [%s] servers or node capacity to assign segment[%s]! Expected Replicants[%d]",
+            tier, segment.getIdentifier(), targetReplicantsInTier
         );
+        return numAssigned;
       }
+      holders.remove(holder);
 
-      holder.getPeon().loadSegment(
-          segment,
-          new LoadPeonCallback()
-          {
-            @Override
-            public void execute()
-            {
-              replicationManager.unregisterReplicantCreation(
-                  tier,
-                  segment.getIdentifier(),
-                  holder.getServer().getHost()
-              );
-            }
-          }
-      );
-
-      ++assignedCount;
-      ++currReplicantsInTier;
-      ++currTotalReplicantsInCluster;
+      final String segmentId = segment.getIdentifier();
+      final String holderHost = holder.getServer().getHost();
+      throttler.registerReplicantCreation(tier, segmentId, holderHost);
+      holder.getPeon().loadSegment(segment, () -> throttler.unregisterReplicantCreation(tier, segmentId, holderHost));
     }
 
-    return assignedCount;
+    return numToAssign;
   }
 
-  private CoordinatorStats drop(
-      final Map<String, Integer> loadStatus,
+  private void drop(
+      final DruidCoordinatorRuntimeParams params,
       final DataSegment segment,
-      final DruidCoordinatorRuntimeParams params
+      final CoordinatorStats stats
   )
   {
-    CoordinatorStats stats = new CoordinatorStats();
+    final DruidCluster druidCluster = params.getDruidCluster();
 
     // Make sure we have enough loaded replicants in the correct tiers in the cluster before doing anything
-    for (Integer leftToLoad : loadStatus.values()) {
-      if (leftToLoad > 0) {
-        return stats;
+    // This enforces that loading is completed before we attempt to drop stuffs as a safety measure
+    for (final Object2IntMap.Entry<String> entry : targetReplicants.object2IntEntrySet()) {
```

Comment:
Please extract this block as a method with boolean result

Reply:
👍